source/GameAction.cpp

```diff
@@ -0,0 +1,228 @@
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "GameAction.h"
+
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "Dialog.h"
+#include "Format.h"
+#include "GameData.h"
+#include "GameEvent.h"
+#include "PlayerInfo.h"
+#include "Random.h"
+
+#include <cstdlib>
+#include <vector>
+
+using namespace std;
+
+
+
+// Construct and Load() at the same time.
+GameAction::GameAction(const DataNode &node, const string &missionName)
+{
+	Load(node, missionName);
+}
+
+
+
+void GameAction::Load(const DataNode &node, const string &missionName)
+{
+	empty = false;
+	for(const DataNode &child : node)
+		LoadSingle(child, missionName);
+}
+
+
+
+// Load a single child at a time, used for streamlining MissionAction::Load.
+void GameAction::LoadSingle(const DataNode &child, const string &missionName)
+{
+	const string &key = child.Token(0);
+	bool hasValue = (child.Size() >= 2);
+	
+	if(key == "log")
+	{
+		bool isSpecial = (child.Size() >= 3);
+		string &text = (isSpecial ?
+			specialLogText[child.Token(1)][child.Token(2)] : logText);
+		Dialog::ParseTextNode(child, isSpecial ? 3 : 1, text);
+	}
+	else if(key == "payment")
+	{
+		if(child.Size() == 1)
+			paymentMultiplier += 150;
+		if(child.Size() >= 2)
+			payment += child.Value(1);
+		if(child.Size() >= 3)
+			paymentMultiplier += child.Value(2);
+	}
+	else if(key == "event" && hasValue)
+	{
+		int minDays = (child.Size() >= 3 ? child.Value(2) : 0);
+		int maxDays = (child.Size() >= 4 ? child.Value(3) : minDays);
+		if(maxDays < minDays)
+			swap(minDays, maxDays);
+		events[GameData::Events().Get(child.Token(1))] = make_pair(minDays, maxDays);
+	}
+	else if(key == "fail")
+	{
+		string toFail = child.Size() >= 2 ? child.Token(1) : missionName;
+		fail.insert(toFail);
+		// Create a GameData reference to this mission name.
+		GameData::Missions().Get(toFail);
+	}
+	else
+		conditions.Add(child);
+}
+
+
+
+void GameAction::Save(DataWriter &out) const
+{
+	if(!logText.empty())
+	{
+		out.Write("log");
+		out.BeginChild();
+		{
+			// Break the text up into paragraphs.
+			for(const string &line : Format::Split(logText, "\n\t"))
+				out.Write(line);
+		}
+		out.EndChild();
+	}
+	for(const auto &it : specialLogText)
+		for(const auto &eit : it.second)
+		{
+			out.Write("log", it.first, eit.first);
+			out.BeginChild();
+			{
+				// Break the text up into paragraphs.
+				for(const string &line : Format::Split(eit.second, "\n\t"))
+					out.Write(line);
+			}
+			out.EndChild();
+		}
+	if(payment)
+		out.Write("payment", payment);
```

Comment:
Should this also write out the paymentMultiplier if there is one? (Probably not, since it gets handled in instantiate, but maybe add a comment here that the paymentMultiplier is never used in dynamic data.)

Reply:
GameAction is basically just ripped out of MissionAction, so if it works in the game right now then it’ll work in this PR. :)